File: Sources/CCUsageCore/Engine/ConfigDir.swift
import Foundation
import CryptoKit

/// Um diretório de configuração do Claude Code — um perfil.
///
/// Guarda a **string crua** exportada como `CLAUDE_CONFIG_DIR`, porque é ela, e
/// não o caminho resolvido, que o Claude Code normaliza em NFC e hasheia no nome
/// do item de chaveiro. Resolver o caminho (tirar um `~`, um `./`, uma barra
/// final) mudaria o hash e apontaria para um item que não existe. Confirmado
/// contra os itens reais no chaveiro em agosto/2026.
public struct ConfigDir: Sendable, Equatable, Hashable, Codable {
    /// A string exatamente como vai para o ambiente. Para o perfil padrão é o
    /// caminho de `~/.claude`; para os outros, o caminho do diretório do grupo.
    public let raw: String

    /// `true` quando é `~/.claude` — o perfil que o `claude` usa sem nenhuma
    /// variável de ambiente.
    ///
    /// Não é cosmético: setar `CLAUDE_CONFIG_DIR=~/.claude` **não** é o mesmo que
    /// não setar — com a variável presente a sessão sobe deslogada. Então o
    /// perfil padrão se distingue por não exportar variável nenhuma, e é isto que
    /// esta marca controla.
    public let isDefault: Bool

    public init(raw: String, isDefault: Bool) {
        self.raw = raw
        self.isDefault = isDefault
    }

    /// O perfil padrão, `~/.claude`. Nunca exporta `CLAUDE_CONFIG_DIR`.
    public static func standard(home: String = NSHomeDirectory()) -> ConfigDir {
        ConfigDir(raw: "\(home)/.claude", isDefault: true)
    }

    /// Um perfil dedicado, num caminho qualquer. Exporta `CLAUDE_CONFIG_DIR`.
    public static func dedicated(_ path: String) -> ConfigDir {
        ConfigDir(raw: path, isDefault: false)
    }

    /// A URL do diretório em si.
    public var url: URL { URL(fileURLWithPath: raw) }

    /// O `.claude.json` deste perfil.
    ///
    /// Assimetria confirmada na fonte: no perfil padrão ele mora **ao lado** do
    /// diretório (`~/.claude.json`, não `~/.claude/.claude.json`); num perfil com
    /// `CLAUDE_CONFIG_DIR` setado, mora **dentro** (`<dir>/.claude.json`).
    public var globalConfigURL: URL {
        isDefault
            ? url.deletingLastPathComponent().appending(path: ".claude.json")
            : url.appending(path: ".claude.json")
    }

    /// A variável de ambiente que uma sessão neste perfil precisa — ou `nil` no
    /// padrão, que não exporta nada.
    public var environmentValue: String? { isDefault ? nil : raw }

    /// O sufixo de 8 hex que o Claude Code deriva de um caminho de perfil.
    ///
    /// `sha256(NFC(raw))[:8]`. Só faz sentido para perfis não-padrão; o padrão
    /// usa o nome de item sem sufixo.
    public var keychainHash: String {
        let normalized = raw.precomposedStringWithCanonicalMapping  // NFC
        let digest = SHA256.hash(data: Data(normalized.utf8))
        return digest.prefix(4).map { String(format: "%02x", $0) }.joined()
    }
}

File: windows/crates/router-core/src/engine/config_dir.rs
//! Um diretório de configuração do Claude Code — um perfil.
//!
//! Guarda a **string crua** exportada como `CLAUDE_CONFIG_DIR` (no macOS ela vira
//! o hash do item de chaveiro; no Windows a credencial é arquivo, então não há
//! hash — mas o formato do `config.json` mantém `{raw, isDefault}` idêntico).

use std::path::PathBuf;

use serde::{Deserialize, Serialize};

#[derive(Clone, PartialEq, Eq, Hash, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConfigDir {
    /// A string exatamente como vai para o ambiente. No perfil padrão é o caminho
    /// de `<home>\.claude`; nos outros, o caminho do diretório do grupo.
    pub raw: String,

    /// `true` quando é o perfil padrão — o que o `claude` usa sem nenhuma
    /// variável de ambiente.
    ///
    /// Não é cosmético: setar `CLAUDE_CONFIG_DIR` para o caminho padrão **não** é
    /// o mesmo que não setar — com a variável presente a sessão sobe deslogada
    /// (confirmado no macOS; a checagem isolada no Windows fica para a
    /// implementação). O padrão se distingue por não exportar variável nenhuma.
    pub is_default: bool,
}

impl ConfigDir {
    /// O perfil padrão, `<home>\.claude`. Nunca exporta `CLAUDE_CONFIG_DIR`.
    pub fn standard(home: &str) -> Self {
        ConfigDir {
            raw: format!("{home}/.claude"),
            is_default: true,
        }
    }

    /// Um perfil dedicado, num caminho qualquer. Exporta `CLAUDE_CONFIG_DIR`.
    pub fn dedicated(path: impl Into<String>) -> Self {
        ConfigDir {
            raw: path.into(),
            is_default: false,
        }
    }

    /// O caminho do diretório em si.
    pub fn path(&self) -> PathBuf {
        PathBuf::from(&self.raw)
    }

    /// O `.claude.json` deste perfil.
    ///
    /// Assimetria confirmada no disco (macOS e Windows): no perfil padrão mora
    /// **ao lado** do diretório (`<home>\.claude.json`, não
    /// `<home>\.claude\.claude.json`); num perfil dedicado, mora **dentro**
    /// (`<dir>\.claude.json`).
    pub fn global_config_path(&self) -> PathBuf {
        let dir = self.path();
        if self.is_default {
            match dir.parent() {
                Some(parent) => parent.join(".claude.json"),
                None => PathBuf::from(".claude.json"),
            }
        } else {
            dir.join(".claude.json")
        }
    }

    /// A variável de ambiente que uma sessão neste perfil precisa — ou `None` no
    /// padrão, que não exporta nada.
    pub fn environment_value(&self) -> Option<&str> {
        if self.is_default {
            None
        } else {
            Some(&self.raw)
        }
    }
}
